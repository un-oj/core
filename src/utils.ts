export { version } from '../package.json' with { type: 'json' };

/** Get the first key of an object. */
export function getFirstKey<T extends Record<never, never>>(obj: T): keyof T {
  return Object.keys(obj)[0] as keyof T;
}

/**
 * Parses a human-readable time string.
 * @param s The string to parse.
 * @returns The time in milliseconds, or undefined if failed.
 */
export function parseTime(s: string): number | undefined {
  const val = Number.parseInt(s);
  if (Number.isNaN(val))
    return;
  switch (s) {
    case `${val} second`:
    case `${val} seconds`:
    case `${val} sec`:
    case `${val} secs`:
      return val * 1000;
  }
}

const MEMORY_UNITS: Record<string, number> = {
  megabyte: 1024 * 1024,
  megabytes: 1024 * 1024,
  MB: 1024 * 1024,
  MiB: 1024 * 1024,
};

/**
 * Parses a human-readable memory size string.
 *
 * @param s The string to parse.
 * @returns The memory size in bytes, or undefined if failed.
 */
export function parseMemory(s: string): number | undefined {
  const [, val, unit] = s.match(/^(\d+) ([a-z]+)$/i) ?? [];
  if (!val || !unit)
    return;
  return Number(val) * MEMORY_UNITS[unit];
}

/**
 * Adds key-value pairs to the provided headers object if they aren't present.
 *
 * @param h The original headers object.
 * @param add The key-value pairs to add.
 * @returns A **new** headers object.
 */
export function addHeaders(h: HeadersInit | undefined, add: Array<[string, string]>): Record<string, string> {
  if (h instanceof Headers)
    h = h.toJSON();
  if (Array.isArray(h))
    h = Object.fromEntries(h);
  else if (h == null)
    h = {};
  else
    h = structuredClone(h);

  for (const item of add) {
    if (!(item[0] in h))
      h[item[0]] = item[1];
  }
  return h;
}

/** General error class for UnOJ. */
export class UnOJError extends Error {}
