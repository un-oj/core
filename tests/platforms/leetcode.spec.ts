import { NotFoundError } from '@un-oj/core';
import LeetCode from '@un-oj/core/platforms/leetcode';
import { describe, expect, it } from 'bun:test';
import { assertProblem } from './utils.ts';

describe('LeetCode platform', () => {
  const lc = new LeetCode();
  const lcCN = new LeetCode({ baseURL: 'https://leetcode.cn' });

  it('should fetch problem', async () => {
    assertProblem(await lc.getProblem('two-sum'));
  });

  it('should fetch problem w/ LeetCode CN', async () => {
    assertProblem(await lcCN.getProblem('two-sum'));
  });

  it('should throw NotFoundError w/ non-existent problem', async () => {
    expect(lc.getProblem('this-problem-does-not-exist')).rejects.toThrow(NotFoundError);
  });
});
