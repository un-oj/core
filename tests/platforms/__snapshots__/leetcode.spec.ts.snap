// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`LeetCode platform should fetch problem 1`] = `
{
  "description": 
"<p>Given an array of integers <code>nums</code>&nbsp;and an integer <code>target</code>, return <em>indices of the two numbers such that they add up to <code>target</code></em>.</p>

<p>You may assume that each input would have <strong><em>exactly</em> one solution</strong>, and you may not use the <em>same</em> element twice.</p>

<p>You can return the answer in any order.</p>

<p>&nbsp;</p>
<p><strong class="example">Example 1:</strong></p>

<pre>
<strong>Input:</strong> nums = [2,7,11,15], target = 9
<strong>Output:</strong> [0,1]
<strong>Explanation:</strong> Because nums[0] + nums[1] == 9, we return [0, 1].
</pre>

<p><strong class="example">Example 2:</strong></p>

<pre>
<strong>Input:</strong> nums = [3,2,4], target = 6
<strong>Output:</strong> [1,2]
</pre>

<p><strong class="example">Example 3:</strong></p>

<pre>
<strong>Input:</strong> nums = [3,3], target = 6
<strong>Output:</strong> [0,1]
</pre>

<p>&nbsp;</p>
<p><strong>Constraints:</strong></p>

<ul>
	<li><code>2 &lt;= nums.length &lt;= 10<sup>4</sup></code></li>
	<li><code>-10<sup>9</sup> &lt;= nums[i] &lt;= 10<sup>9</sup></code></li>
	<li><code>-10<sup>9</sup> &lt;= target &lt;= 10<sup>9</sup></code></li>
	<li><strong>Only one valid answer exists.</strong></li>
</ul>

<p>&nbsp;</p>
<strong>Follow-up:&nbsp;</strong>Can you come up with an algorithm that is less than <code>O(n<sup>2</sup>)</code><font face="monospace">&nbsp;</font>time complexity?"
,
  "difficulty": "Easy",
  "id": "two-sum",
  "link": "https://leetcode.com/problems/two-sum/",
  "memoryLimit": undefined,
  "samples": [
    {
      "hint": "Because nums[0] + nums[1] == 9, we return [0, 1].",
      "input": "nums = [2,7,11,15], target = 9",
      "output": "[0,1]",
    },
    {
      "input": "nums = [3,2,4], target = 6",
      "output": "[1,2]",
    },
    {
      "input": "nums = [3,3], target = 6",
      "output": "[0,1]",
    },
  ],
  "tags": [
    "array",
    "hash-table",
  ],
  "timeLimit": undefined,
  "title": "Two Sum",
  "type": "traditional",
}
`;

exports[`LeetCode platform should fetch problem w/ LeetCode CN 1`] = `
{
  "description": 
"<p>给定一个整数数组 <code>nums</code>&nbsp;和一个整数目标值 <code>target</code>，请你在该数组中找出 <strong>和为目标值 </strong><em><code>target</code></em>&nbsp; 的那&nbsp;<strong>两个</strong>&nbsp;整数，并返回它们的数组下标。</p>

<p>你可以假设每种输入只会对应一个答案，并且你不能使用两次相同的元素。</p>

<p>你可以按任意顺序返回答案。</p>

<p>&nbsp;</p>

<p><strong class="example">示例 1：</strong></p>

<pre>
<strong>输入：</strong>nums = [2,7,11,15], target = 9
<strong>输出：</strong>[0,1]
<strong>解释：</strong>因为 nums[0] + nums[1] == 9 ，返回 [0, 1] 。
</pre>

<p><strong class="example">示例 2：</strong></p>

<pre>
<strong>输入：</strong>nums = [3,2,4], target = 6
<strong>输出：</strong>[1,2]
</pre>

<p><strong class="example">示例 3：</strong></p>

<pre>
<strong>输入：</strong>nums = [3,3], target = 6
<strong>输出：</strong>[0,1]
</pre>

<p>&nbsp;</p>

<p><strong>提示：</strong></p>

<ul>
	<li><code>2 &lt;= nums.length &lt;= 10<sup>4</sup></code></li>
	<li><code>-10<sup>9</sup> &lt;= nums[i] &lt;= 10<sup>9</sup></code></li>
	<li><code>-10<sup>9</sup> &lt;= target &lt;= 10<sup>9</sup></code></li>
	<li><strong>只会存在一个有效答案</strong></li>
</ul>

<p>&nbsp;</p>

<p><strong>进阶：</strong>你可以想出一个时间复杂度小于 <code>O(n<sup>2</sup>)</code> 的算法吗？</p>
"
,
  "difficulty": "Easy",
  "id": "two-sum",
  "link": "https://leetcode.cn/problems/two-sum/",
  "memoryLimit": undefined,
  "samples": [
    {
      "hint": "因为 nums[0] + nums[1] == 9 ，返回 [0, 1] 。",
      "input": "nums = [2,7,11,15], target = 9",
      "output": "[0,1]",
    },
    {
      "input": "nums = [3,2,4], target = 6",
      "output": "[1,2]",
    },
    {
      "input": "nums = [3,3], target = 6",
      "output": "[0,1]",
    },
  ],
  "tags": [
    "array",
    "hash-table",
  ],
  "timeLimit": undefined,
  "title": "两数之和",
  "type": "traditional",
}
`;
